pycbc/inference/evidence.py

```diff
@@ -0,0 +1,465 @@
+    # we can sacrifice precision there, rather than near beta -> 1
+    # where we need high precision. Use trapezoid rule on that point.
+    log_evidence = integrate.simps(average_logl, betas, even="last")
+
+   # Estimating the standard error of the thermodynamic integration
+    # evidence.
+    # See https://www.sciencedirect.com/science/article/pii/S0022249617302651
+    # Eqs  (36) and (37)
+    TI_vec = numpy.zeros(len(log_likelihood[0]))
+    for i in range(len(log_likelihood[0])):
+        log_likelihood_samps = [log_likelihood[x][i] for x in range(len(betas))]
+        TI_vec[i] = integrate.simps(log_likelihood_samps, betas,
+                                    even="last")
+
+    mcmc_std = numpy.std(TI_vec) / numpy.sqrt(float(len(log_likelihood[0])))
+
+    return log_evidence, mcmc_std
+
+
+def stepping_stone_algorithm(log_likelihood, betas):
+    """
+    Returns the evidence of the model via stepping stone
+    algorithm, and an estimated standard deviation for the
+    evidence.
+
+    Parameters
+    ----------
+
+    log_likelihood : 3d array of shape (betas, walker, iteration)
+                The log likelihood for each temperature separated
+                by temperature, walker, and iteration.
+
+    betas          : 1d array
+                     The inverse temperatures used in the MCMC.
+
+    Returns
+    ----------
+    log_evidence    : float
+                    Estimation of the log of the evidence.
+    mcmc_std: float
+                    The standard deviation of the evidence estimate
+                    based on the deviation of log likelihoods for
+                    each temperature.
+    """
+    # Reverse order sort
+    order = numpy.argsort(betas)[::-1]
+    betas = betas[order]
+    log_likelihood = log_likelihood[order]
+
+    # Assume log likelihood is given in shape of beta,
+    # walker, iteration.
+    log_likelihood = numpy.reshape(log_likelihood, (len(betas),
+                                                    len(log_likelihood[0].flatten())))
+
+    log_rk_pb = numpy.zeros(len(betas) - 1)
+    for i in range(len(betas) - 1):
+        betas_diff = betas[i] - betas[i+1]
+        # Max log likelihood for beta [i+1]
+        max_logl_pb = numpy.max(log_likelihood[i+1])
+        val_1 = betas_diff * max_logl_pb
+        val_2 = numpy.log(numpy.average(numpy.exp(betas_diff * (log_likelihood[i+1] - max_logl_pb))))
+
+        log_rk_pb[i] = val_1 + val_2
+    log_rk = numpy.sum(log_rk_pb)
+    log_evidence = log_rk
+
+    mcmc_std = 0
+    for i in range(len(betas) - 1):
+        betas_diff = betas[i] - betas[i+1]
+
+        pre_fact = numpy.exp((betas_diff * log_likelihood[i+1]) - log_rk_pb[i]) - 1.0
+        val_2 = numpy.sum(pre_fact ** 2)
+
+        mcmc_std += val_2
+
+    mcmc_std /= float(len(log_likelihood[0])) ** 2.0
+    mcmc_std = numpy.sqrt(mcmc_std)
+
+    return log_evidence, mcmc_std
+
+def calculate_log_bayes_factor(log_ev_1, log_ev_2):
```

Comment:
Does this need it's own function to do a simple difference?

Reply:
Probably not. I did want to have some bayes factor convenience functions though. I can push them to another module.